source/slang/syntax.cpp

```diff
@@ -940,6 +940,155 @@ void Type::accept(IValVisitor* visitor, void* extra)
         return this;
     }
 
+    // AssocTypeDeclRefType
+
+    String AssocTypeDeclRefType::ToString()
+    {
+        // TODO: what is appropriate here?
+        return "<AssocType>";
+    }
+
+    bool AssocTypeDeclRefType::EqualsImpl(Type * type)
+    {
+        if (auto assocTypeDeclRefType = type->As<AssocTypeDeclRefType>())
+        {
+            return declRef.Equals(assocTypeDeclRefType->declRef);
+        }
+        return false;
+    }
+
+    RefPtr<Val> AssocTypeDeclRefType::SubstituteImpl(Substitutions* subst, int* ioDiff)
+    {
+        if (!sourceType)
+            return this;
+        if (auto parentDeclRefType = sourceType->As<DeclRefType>())
+        {
+            auto parentDeclRef = parentDeclRefType->declRef;
+            DeclRef<AggTypeDecl> newParentDeclRef = parentDeclRef.As<AggTypeDecl>();
+            // search for a substitution that might apply to us
+            for (auto s = subst; s; s = s->outer.Ptr())
+            {
+                // the generic decl associated with the substitution list must be
+                // the generic decl that declared this parameter
+                auto genericDecl = s->genericDecl;
+                if (genericDecl != parentDeclRef.getDecl()->ParentDecl)
+                    continue;
+                int index = 0;
+                for (auto m : genericDecl->Members)
+                {
+                    if (m.Ptr() == parentDeclRef.getDecl())
+                    {
+                        // We've found it, so return the corresponding specialization argument
+                        (*ioDiff)++;
+                        if (auto declRef = s->args[index].As<DeclRefType>())
+                        {
+                            newParentDeclRef = (*declRef).declRef.As<AggTypeDecl>();
+                            goto searchEnd;
+                        }
+                    }
+                    else if (auto typeParam = m.As<GenericTypeParamDecl>())
+                    {
+                        index++;
+                    }
+                    else if (auto valParam = m.As<GenericValueParamDecl>())
+                    {
+                        index++;
+                    }
+                    else
+                    {
+                    }
+                }
+            }
+        searchEnd:
+            if (newParentDeclRef)
+            {
+                Decl* targetTypeDecl = nullptr;
+                if (newParentDeclRef.getDecl()->memberDictionary.TryGetValue(this->GetDeclRef().decl->getName(), targetTypeDecl))
```

Comment:
Ack! I really don't like the idea that we perform by-name member lookup here.

It seems like the right logic that should flow here is that if we used to have a reference to `SomeInterfaceType::SomeAssociatedType` but now we are trying to form `SomeConcreteType::SomeAssociatedType` then we need to find the conformance that tells us `SomeConcreteType : SomeInterfaceType` (ideally that conformance is a reference to either a `GenericTypeConstraintDecl` or an `InheritanceDecl` and we don't care which - those two probably need to share a base class), then the declaration that witnesses the conformance will have a pre-built mapping from the decls in the interface to their realization in the concrete type, and we can just use *that* for lookup and not have to deal with any of the vagaries of overload resolution, inheritance, etc.